src/ui/RadarMarker.ts

```diff
@@ -0,0 +1,126 @@
+import maplibregl from 'maplibre-gl';
+
+import Http from '../http';
+import Logger from '../logger';
+
+import type RadarMap from './RadarMap';
+
+import type { RadarMarkerImage, RadarMarkerOptions } from '../types';
+
+const defaultMarkerOptions: maplibregl.MarkerOptions = {
+  color: '#000257',
+};
+
+const fileExtensionToContentType: Record<string, string> = {
+  svg: 'image/svg+xml',
+  png: 'image/png',
+};
+
+const getFileExtension = (file: string): string => {
+  return file.split('.').pop() || '';
+}
+
+const createImageElement = (options: RadarMarkerImage) => {
+  const element = document.createElement('img');
+  element.src = options.url!;
+
+  if (options.width) {
+    element.width = options.width;
+  }
+  if (options.height) {
+    element.height = options.height;
+  }
+  if (!options.width && !options.height) {
+    element.style.maxWidth = '64px';
+    element.style.maxHeight = '64px';
+  }
+  return element;
+}
+class RadarMarker extends maplibregl.Marker {
+  _map!: RadarMap;
+  _image?: RadarMarkerImage;
+
+  constructor(markerOptions: RadarMarkerOptions) {
+    const maplibreOptions: maplibregl.MarkerOptions = Object.assign({}, defaultMarkerOptions);
+
+    if (markerOptions.color) {
+      maplibreOptions.color = markerOptions.color;
+    }
+    if (markerOptions.element) {
+      maplibreOptions.element = markerOptions.element;
+    }
+    if (markerOptions.scale) {
+      maplibreOptions.scale = markerOptions.scale;
+    }
+
+    super(maplibreOptions);
+
+    if (markerOptions.image) {
+      this._image = markerOptions.image;
+
+      const originalElement = this._element.cloneNode(true);
+      this._element.childNodes.forEach((child) => {
+        child.remove();
+      });
+
+      const onSuccess = (blob: Blob) => {
+        const customMarkerUrl = URL.createObjectURL(blob);
+        this._element.replaceChildren(createImageElement({ url: customMarkerUrl }));
+      };
+
+      const onError = (err: any) => {
+        Logger.error(`Could not get custom marker: ${err.message} - falling back to default marker`);
+        this._element.replaceChildren(...originalElement.childNodes);
+      }
+
+      if (markerOptions.image.url) {
+        fetch(markerOptions.image.url)
+          .then(res => {
+            if (res.status === 200) {
+              res.blob()
+                .then(onSuccess)
+                .catch(onError);
+            } else {
+              onError(new Error(res.statusText));
+            }
+          })
+          .catch(onError)
+      }
```

Comment:
not the prettiest way to handle this but we can't use `async await` in a constructor. if `markerOptions.image.url` exists, technically we can just do something like `img.src = markerOptions.image.url` but writing this way allows the marker to fallback to the default if there is an error with the provided url.